### PR TITLE
Adding support for Oracle Cloud Infrastructure (OCI) Object Store as Delta Storage

### PR DESCRIPTION
**Pull request by vibhaska**
Adding support for Oracle Cloud Infrastructure (OCI) Object Store as Delta Storage by introducing OCILogStore.

Regarding [Storage configuration](https://docs.delta.io/latest/delta-storage.html) page in Delta Documentation, I request following changes mentioned [here](https://docs.google.com/document/d/1DJvRAuUWUov5kepAQb176uSUsdlgU2MxaGBDBYiRGsg/edit?usp=sharing).

### CODE REVIEW

**Comment by pranavanand** [comment]
Thanks for submitting this PR! We are currently coming up with a process for contributing third-party support and will let you know once that's ready.

**Comment by ksranga** [comment]
Can this be ported to delta 0.6.x version so its available for spark 2.x as well?

**Comment by vibhaska** [comment]
@ksranga Yes.

I guess only one line change is required: Remove line number 268 from LogStoreSuite.scala. If you want I can raise another PR for old branch after testing and verification.

File: README.md [hunk 1]
```diff
@@ -88,7 +88,7 @@ Delta Lake ACID guarantees are predicated on the atomicity and durability guaran
...
+Given that storage systems do not necessarily provide all of these guarantees out-of-the-box, Delta Lake transactional operations typically go through the [LogStore API](https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/storage/LogStore.scala) instead of accessing the storage system directly. We can plug in custom `LogStore` implementations in order to provide the above guarantees for different storage systems. Delta Lake has built-in `LogStore` implementations for HDFS, Amazon S3, Azure and OCI(Oracle Cloud Infrastructure) storage services. Please see [Delta Lake Storage Configuration](https://docs.delta.io/latest/delta-storage.html) for more details. If you are interested in adding a custom `LogStore` implementation for your storage system, you can start discussions in the community mailing group.
```
Comment (mikedillion) [inline README.md:91]: Missing a space here

```diff
- OCI(Oracle Cloud Infrastructure)
+ OCI (Oracle Cloud Infrastructure)
```

**Comment by vibhaska** [comment]
> Thanks for submitting this PR! We are currently coming up with a process for contributing third-party support and will let you know once that's ready.

@pranavanand can you please share any further update.

**Comment by ddraj** [comment]
Echoing @vibhaska's question .. Hi @pranavanand, any chance the PR can be merged? Could you please let us know if anything is pending on the PR that we should address?

**Comment by tdas** [comment]
Hey @ddraj 

Our apologies for sitting on the PR for so long. We wanted to figure out what is the best we integrate slightly less battle-tested features into our repo, and we finally figured it out. We are currently in the process of converting our repo to a [multi-module SBT project](https://github.com/delta-io/delta/pull/644) so that we can add other sub-modules (mapping to other maven release artifacts beyond delta-core). Then we can add a contribs module and then accept such new and experimental features.

Are you still interested in contributing to this feature?

**Comment by ddraj** [comment]
> Hey @ddraj
> 
> Our apologies for sitting on the PR for so long. We wanted to figure out what is the best we integrate slightly less battle-tested features into our repo, and we finally figured it out. We are currently in the process of converting our repo to a [multi-module SBT project](https://github.com/delta-io/delta/pull/644) so that we can add other sub-modules (mapping to other maven release artifacts beyond delta-core). Then we can add a contribs module and then accept such new and experimental features.
> 
> Are you still interested in contributing to this feature?

Thanks @tdas. Yes we are still interested. Please let us know if we need to do anything specific.

**Comment by tdas** [comment]
So the refactor to multi-module has been merged. Lets' work towards getting this in. Here are the next steps.

1. Please define a new module using the same pattern as the `core` module in the build.sbt file.
    - Does the IBM support need additional maven dependencies? I am guessing not. In that case, we can make the module name generic like "delta-contribs" (that is, not specific to IBM). Other such zero-dependency contributions can go in that project as well. Otherwise, lets put it in a specifically named module like "delta-contribs-ibm"
   
2. Move the new log store code inside the `<new-module-dir>/src/main/<package>/...` and the test code in a separate file in `<new-module-dir>/src/test/<package>/...`.

Let me know if you need help to get it working.

**Comment by vibhaska** [comment]
@tdas I have updated the changes as suggested. Kindly check if this is what you wanted.

Couple of points where I need some guidance:
1. Will there be any root project, and jars from same will be published? Or will a separate jars be published from this module. I am not clear how.
2. 2. I am also not clear why CI failed. Is this due to multi module nature?

**Comment by tdas** [comment]
These are good questions. My apologies for the delay, we have been trying to line up a few more features like a more stable [LogStore API](https://github.com/delta-io/delta/commit/b7cb203dd23a9a225fbcc414bc075916bd32a817) that LogStore builder like you guys can use. However, there are a few pieces that we still need to get it for the public API to be easily usable. So we can continue working using the existing API, which we will continue supporting for at least the 1.x releases.

Now back to the question. We will be releasing a separate Maven artifact for each subproject. The refactoring you have done in the sbt looks good at a high-level. I will leave more detailed comments on them soon.

**Comment by tdas** [comment]
Somehow the sbt download failed. Could be a random error. Can you run the tests once again. An easy to rerun tests is to simply push an empty commit to this branch.

File: build.sbt [hunk 2]
```diff
@@ -95,6 +95,79 @@ lazy val core = (project in file("core"))
...
+  .dependsOn(core % "compile->compile;test->test")
+  .enablePlugins(GenJavadocPlugin, JavaUnidocPlugin, ScalaUnidocPlugin)
+  .settings (
+    name := "delta-contribs-oci",
+    commonSettings,
+    scalaStyleSettings,
+    // mimaSettings,
+    unidocSettings,
+    releaseSettings,
+    libraryDependencies ++= Seq(
```
Comment (tdas) [inline build.sbt:109]: do we need to add these dependencies? I think this project automatically gets the dependencies from the core module

File: build.sbt [hunk 3]
```diff
@@ -95,6 +95,79 @@ lazy val core = (project in file("core"))
...
+lazy val oci = (project in file("delta-contribs-oci"))
+  .aggregate(core)
+  .dependsOn(core % "compile->compile;test->test")
+  .enablePlugins(GenJavadocPlugin, JavaUnidocPlugin, ScalaUnidocPlugin)
+  .settings (
+    name := "delta-contribs-oci",
+    commonSettings,
+    scalaStyleSettings,
+    // mimaSettings,
+    unidocSettings,
```
Comment (tdas) [inline build.sbt:107]: i dont think we need unidoc. there is no public in this project yet.

**Comment by tdas** [comment]
The community has already created the contrib submodule! So all you have to do merge with the master and add your LogStore to the existing directories. No need to fix anything in the build.sbt!

**Comment by tdas** [comment]
@vibhaska @ddraj if you have time to update the PR, please do so quickly as we preparing to release 1.0 soon. Alternatively, if you are okay, I can do the slight refactoring as well (you still get commit authorship).

**Comment by ddraj** [comment]
@tdas coincidentally @vibhaska is on vacation this week and the next. So we are not able to update the PR :( I'll happily take your offer for the slight updates. Thanks a lot @tdas

**Comment by tdas** [comment]
@ddraj all right! I will do it. Do you mind if I change the name OracleCILogstore to make it more obvious which cloud it is? Similar to our naming of IBMCOSLogStore and AzureLogStore.

**Comment by vibhaska** [comment]
Sorry for the delayed response. I have checked-in refactoring changes. I have also updated the file name, as suggested.

@tdas Thanks for offering help with the changes. Kindly review and please suggest if any more changes are required.

@ddraj Thanks for your support.

**Comment by ddraj** [comment]
@tdas any chance we can name it as OracleCloudLogStore. I think that would fully convey the purpose of this implementation.

**Comment by tdas** [comment]
Sure. As long as there won't be any other variation of LogStore needed on the Oracle Cloud, maybe for any other storage product in Oracle Cloud.

**Comment by tdas** [comment]
Alright!

**Comment by vibhaska** [comment]
@tdas @ddraj Updated the file name, as suggested.

**Comment by vibhaska** [comment]
@tdas Kindly suggest if any further changes are required.

**Comment by ddraj** [comment]
@tdas echoing what @vibhaska asked. Is there anything you need from us?

**Comment by ddraj** [comment]
Hi @tdas gentle ping :)

**Pull request closed without merging**